gcc/rust/ast/rust-ast.cc

```diff
@@ -4223,26 +4223,32 @@ Attribute::is_parsed_to_meta_item () const
 }
 
 void
-BlockExpr::strip_tail_expr ()
+BlockExpr::try_convert_last_stmt ()
 {
-  if (expr)
+  rust_assert (expr == nullptr);
+  // HACK: try to turn the last statement into a tail expression
+  if (statements.size () && statements.back ()->is_expr ())
     {
-      expr = nullptr;
+      auto &stmt = static_cast<ExprStmt &> (*statements.back ());
```

Comment:
Just note it is undefined behaviour to take a reference to a vector element then mutate the vector after then use the same reference again.

You are ok here because you don't use stmt after calling statements.pop_back but if you did it would turn into garbage. I had a horrible issue in the type system where I had a reference to the current context then called query_type which might then change the current context by resolving another function which turned my reference into garbage this was really confusing to track down and it was fixed simply by removing the reference and taking a copy.

Nothing to change here but something to be aware of.